xtra-libp2p/src/endpoint.rs

```diff
@@ -737,3 +742,60 @@ pub struct ListenAddressRemoved {
 impl xtra::Message for ListenAddressRemoved {
     type Result = ();
 }
+
+#[cfg(test)]
```

Comment:
Perhaps I should have been more clear about tests - I meant a test that actually showcases that properly configured failure counter will actually result dropping the connection after X tries.

Reply:
Good call! Once I added a test I noticed that just using a block for the async code is not enough! Bit surprised that the compiler did not complain about this 😅
We actually have to await the result of opening the substream! 
I am not super happy with the second internal function, but doing this with an async block was somewhat messy because of moving what we need from `self`

Reply:
yeah, it's not too pretty, but alternatives are not prettier... I've added a small fixup commit that makes it (subjectively) a tiny bit nicer (and cleaned the test). Feel free to squash it or drop it if you strongly disagree :)